Pr/85 팔로우팔로잉 기능 구현

200wonman opened the pull request:
e.g. resolves #85


# 해결하려는 문제가 무엇인가요?

- 유저간의 팔로우/언팔로우 기능을 구현합니다.

# 어떻게 해결했나요?

- 팔로우/언팔로우 기능을 구현하여 사용자가 다른 사용자를 팔로우하거나 언팔로우할 수 있도록 했습니다.
- 파라미터 isFollow 값을 true/false 활용하여 팔로우 또는 언팔로우 동작을 결정합니다.
- 도메인 이벤트를 사용하여 팔로우/언팔로우 시 관련 로직을 처리하도록 했습니다.

# 어떤 부분에 집중하여 리뷰해야 할까요?

- 팔로우/언팔로우 기능이 정상적으로 동작하는지 확인해 주세요.
- 서비스에서 팔로우/언팔로우의 예외 처리 로직이 적절하게 구현되어 있는지 검토해 주세요.
- 그밖에 계층간 역할분담이 적절하게 분담이 되어있는지 확인해 주세요.

## 참고 자료

- 팔로우/언팔로우 기능의 tec-spec

---

# RCA 룰

r: 꼭 반영해 주세요. 적극적으로 고려해 주세요. (Request changes)  
c: 웬만하면 반영해 주세요. (Comment)  
a: 반영해도 좋고 넘어가도 좋습니다. 그냥 사소한 의견입니다. (Approve)

200wonman committed:
feat : 팔로우, 언팔로우 코드구현

200wonman committed:
feat : 팔로우, 언팔로우 응답코드 수정

200wonman committed:
Merge remote-tracking branch 'origin/main' into pr/85-팔로우팔로잉-기능-구현

200wonman committed:
feat : 주석삭제, todo추가

Whale0928 commented on src/main/java/app/bottlenote/follow/domain/constant/FollowStatus.java:
```diff
@@ -0,0 +1,20 @@
+		return FollowStatus.valueOf(source.toUpperCase());
+	}
+
+}
```
상태명을 나타내는 변수도 추가되면 가독성이 높아질것 같아요.
`BLOCK("차단된 상태")`

200wonman replied:
넵 추가해보겠습니닷

Whale0928 commented on src/main/java/app/bottlenote/follow/dto/FollowUpdateRequest.java:
```diff
@@ -0,0 +1,16 @@
+	@NotNull(message = "팔로우 할 유저의 아이디가 없습니다.")
+	Long followerUserId,
+
+	@NotNull(message = "isFollow 값이 없습니다.")
```
`isFollow`가 어떤 종류의 값인지 다른 곳은 모르지 않을까요 ? 
좀더 명확하게 `isFollow 값이 없습니다 (true:팔로우 , false:언팔로우`)

200wonman replied:
enum으로 여러 상태들을 관리하면서 동시에 메세지에 어떤 값들이 와야하는지 알려주도록 하겠습니다~🥃

Whale0928 commented on src/main/java/app/bottlenote/follow/dto/FollowUpdateRequest.java:
```diff
@@ -0,0 +1,16 @@
+	Long followerUserId,
+
+	@NotNull(message = "isFollow 값이 없습니다.")
+	boolean isFollow
```
래퍼 클래스를 사용해야할것 같네요 
그리고 위에서 선언한 `FollowStatus`를 안쓰는 이유가 있을까요? 보기에는 사용해도 좋을것 같은데.

200wonman replied:
넵!✨

Whale0928 commented on src/main/java/app/bottlenote/follow/dto/FollowUpdateResponse.java:
```diff
@@ -0,0 +1,26 @@
+	@Getter
+	public enum Message {
+		FOLLOW_SUCCESS("성공적으로 팔로우 처리했습니다."),
+		UNFOLLOW_SUCCESS("성공적으로 언팔로우 처리했습니다.");
```
`성공적으로 언팔로우 처리했습니다` -> `성공적으로 팔로우 해제 처리했습니다`
가 좀더 자연스럽지 않을까요?

200wonman replied:
오 좋은 문장 감사합니닷

Whale0928 commented on src/main/java/app/bottlenote/follow/exception/FollowExceptionCode.java:
```diff
@@ -0,0 +1,31 @@
+
+	CANNOT_FOLLOW_SELF(HttpStatus.BAD_REQUEST, "자기 자신을 팔로우할 수 없습니다."),
+	ALREADY_FOLLOWING(HttpStatus.BAD_REQUEST, "이미 팔로우한 사용자입니다."),
+	ALREADY_UNFOLLOWING(HttpStatus.BAD_REQUEST, "이미 언팔로우한 사용자입니다."),
```
HTTP Method가 모두 Bad Request가 맞을까요 ? 다른 메소드가 뭐가 있는지 확인해보죠

200wonman replied:
httpstatus 를 보며 적재적소하게 변경해보겠습니닷

Whale0928 commented on src/main/java/app/bottlenote/follow/service/FollowCommandService.java:
```diff
@@ -0,0 +1,69 @@
+		User user = userRepository.findById(userId)
+			.orElseThrow(() -> new UserException(UserExceptionCode.USER_NOT_FOUND));
+		User followUser = userRepository.findById(followUserId)
+			.orElseThrow(() -> new FollowException(FollowExceptionCode.FOLLOW_NOT_FOUND));
```
적절한 예외 처리 분리 책임  👍

200wonman replied:
👏

Whale0928 commented on src/main/java/app/bottlenote/follow/service/FollowCommandService.java:
```diff
@@ -0,0 +1,69 @@
+				.orElseThrow(() -> new FollowException(FollowExceptionCode.ALREADY_UNFOLLOWING));
+			followRepository.delete(follow);
+			message = FollowUpdateResponse.Message.UNFOLLOW_SUCCESS;
+		}
```
이 방식은 매우 부적절해보입니다.
하나의 메소드에서 팔로우 / 언 팔로우를 모두 수행하고 있는데 
Request에서 Enum을 받아 처리하면 if문 자체가 필요없어 보이네요

또한 `delete`를 호출에 삭제 하는게 맞나요? 
DB에서 값을 삭제하면 언팔로우라는 값은 팔로우 삭제로 적용됩니다.
이 부분은 다시 한번 생각해보세요.

200wonman replied:
follow table에 satus 컬럼을 추가하였으니 상태변경을 위주로코드를 수정하겠습니닷!

Whale0928 reviewed (commented):
동일한 엔드포인트로 구현하려다 보니 붎필요한 복잡성이 많아진것 같습니다.

`@Put` , `@Delete` 으로 팔로우 언팔로우를 분리하는게 좋은 선택이 될수도 있겟네요.

서비스 로직에서 if - else 로직의 경우 개선이 되어야 approve 가능합니다. 🏑

200wonman committed:
Merge remote-tracking branch 'origin/main' into pr/85-팔로우팔로잉-기능-구현

200wonman committed:
refactor : status 컬럼 추가에 따른 팔로우,언팔로우 기능 수정
- 예외케이스를 좀 더 상세히 만졌습니다.
- 너무 과한 if문 사용을 지양했습니다.
- db follow table 에 status 컬럼 추가하여 follow의 status를 활용하여 구현했습니다.

200wonman committed:
Merge remote-tracking branch 'origin/main' into pr/85-팔로우팔로잉-기능-구현

Whale0928 commented on src/main/java/app/bottlenote/follow/domain/Follow.java:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	public Follow updateStatus(FollowStatus followStatus) {
+		if (this.status == followStatus) {
```
동일한 상태로 업데이트한다고 해서 예외처리가 필요하진 않은것 같아요 
1. 동일한 요청이 실수로 동시에 두번 들어올 경우에는?

200wonman replied:
따닥일때 업데이트를 수행하도록 변경하겠습니당 🍹

Whale0928 commented on src/main/java/app/bottlenote/follow/domain/constant/FollowStatus.java:
```diff
@@ -0,0 +1,28 @@
+		if (followStatus == null || followStatus.isEmpty()) {
+			throw new FollowException(FollowExceptionCode.STATUS_NOT_FOUND);
+		}
+		return FollowStatus.valueOf(followStatus.toUpperCase());
```
다른곳을 체크해서 예외시에 어떤식으로 반환해야지 우리가 원하는 예외메시지가 뜰지 체크해봐도 좋을것같아요. 지금은 그냥 전역예외처리 대상이네요

200wonman replied:
이거는 다른 코드 참고해서 했는데, 제가 커스텀한 
VAILD_NOT_STATUS(HttpStatus.BAD_REQUEST, "유효하지 않은 값입니다. FOLLOWING, UNFOLLOW 중 하나여야 합니다."),
예외를 .orElseThrow(() -> new FollowException(FollowExceptionCode.VAILD_NOT_STATUS)); 로 return해도
글로벌이 먼저 잡히더라구요 ㅠ
오히려 .orElse(null);로 하면 null 예외가 잡혀서 첨부된 이미지처럼 저희 원하는 에러메세지에 근사치로 나오지만..
null 익셉션으로 오해의 여지가 생길거 같기도하고 고민입니다. 🍺
<img width="936" alt="스크린샷 2024-05-24 오전 1 00 20" src="https://github.com/bottle-note/bottle-note-api-server/assets/97773895/82c1340e-d220-4863-b862-f4de5f712acd">

Whale0928 commented on src/main/java/app/bottlenote/follow/dto/FollowUpdateResponse.java:
```diff
@@ -0,0 +1,26 @@
+import lombok.Getter;
+
+@Getter
+public class FollowUpdateResponse {
```
대상 유저의 이름 + 썸네일 이미지등을 추가해도 활용여지가 있을것 같네요.

200wonman replied:
추가하겠습니닷 🍻

Whale0928 commented on src/main/java/app/bottlenote/follow/exception/FollowExceptionCode.java:
```diff
@@ -0,0 +1,33 @@
+	ALREADY_FOLLOWING(HttpStatus.CONFLICT, "이미 팔로우한 사용자입니다."),
+	ALREADY_UNFOLLOWING(HttpStatus.CONFLICT, "이미 언팔로우한 사용자입니다."),
+	FOLLOW_NOT_FOUND(HttpStatus.NOT_FOUND, "팔로우할 대상을 찾을 수 없습니다."),
+	CANNOT_UNFOLLOW(HttpStatus.BAD_REQUEST, "팔로우 하지않아 언팔로우 할 수 없습니다."),
```
팔로우하지 않은 상태로 언팔로우하면 그냥 아무 이벤트가 없으면 별 문제 없을것같은데 굳이 예외로 처리해야할까요?

200wonman replied:
불필요한 예외 제거하겠슴니당🧉

Whale0928 commented on src/main/java/app/bottlenote/follow/exception/FollowExceptionCode.java:
```diff
@@ -0,0 +1,33 @@
+	ALREADY_UNFOLLOWING(HttpStatus.CONFLICT, "이미 언팔로우한 사용자입니다."),
+	FOLLOW_NOT_FOUND(HttpStatus.NOT_FOUND, "팔로우할 대상을 찾을 수 없습니다."),
+	CANNOT_UNFOLLOW(HttpStatus.BAD_REQUEST, "팔로우 하지않아 언팔로우 할 수 없습니다."),
+	STATUS_NOT_FOUND (HttpStatus.BAD_REQUEST,"상태가 없습니다 : FOLLOWING, UNFOLLOW, BLOCK, HIDDEN");
```
`상태가 없습니다` 보다 어떤 상태인지 말해주면 더 좋을것같고 (팔로인 상태 )

200wonman replied:
넵~🥃

Whale0928 commented on src/main/java/app/bottlenote/follow/service/FollowCommandService.java:
```diff
@@ -0,0 +1,63 @@
+		Follow follow = followRepository.findByUserIdAndFollowUserId(userId, followUserId)
+			.orElseGet(() -> {
+
+				// 유저와 팔로우유저의 관계가 아무것도 없을때, 언팔로우가 오는것을 방지
```
이거는 위에서 말했듯이 없어도 무의미한것 같아요.

200wonman replied:
유저와 팔로우유저의 관계가 아무것도 없을때, 언팔로우가 오는것을 방지 하지않겠습니당 🍾

Whale0928 commented on src/main/java/app/bottlenote/follow/repository/FollowCommandRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+@Repository
+public interface FollowCommandRepository extends JpaRepository<Follow, Long> {
+	Optional<Follow> findByUserIdAndFollowUserId(Long userId, Long followUserId);
```
fetch join으로 유저와 팔로우 유저를 같이 가져와야 할것 같네요

200wonman replied:
넵 해보겠습니닷 🥤

Whale0928 commented on src/main/java/app/bottlenote/follow/service/FollowCommandService.java:
```diff
@@ -0,0 +1,63 @@
+		followRepository.save(follow);
+
+		return FollowUpdateResponse.builder()
+			.message(request.status() == FollowStatus.FOLLOWING ?
```
메시지 생성시에는 생성된 `follow`에서 상태값만 전달해주고 . 빌더 메소드 내부에서 조절하는것도 좋을것같네요.

200wonman replied:
옙!🍹

Whale0928 reviewed (changes requested):
몇몇 수정사항이 있습니다 . 코멘트를 확인해주세요.

200wonman committed:
Merge remote-tracking branch 'origin/pr/89-presinged-url-기능-구현' into pr/85-팔로우팔로잉-기능-구현

200wonman committed:
refactor : SecurityContextUtil로 userId가져오기 적용

200wonman committed:
refactor : pr의견 반영 수정
- 불필요한 예외제거
- 패치조인 사용
- 응답에 닉네임, 유저프로필이미지 추가

Whale0928 reviewed (dismissed):
1. 그래들 클린 상태에서 빌드와 실행 여부를 모두 검토 후 .
2. 테스트 성공 여부
3. 상위 브랜치 pull받아서 동기화 이후 다시 푸시 후 병합 

위 3개 확인 후 병합 처리 해도 될듯합니다.

200wonman committed:
Merge remote-tracking branch 'origin/pr/89-presinged-url-기능-구현' into pr/85-팔로우팔로잉-기능-구현

200wonman committed:
feat : test code, restDoc 작성
 refactor : follow의 requset,service에서 followerUserId -> followUserId로 수정

200wonman committed:
feat : restdoc 경로수정, 팔로워 챕터 추가
